### PR TITLE
VLM Pipeline (Intern,LLava)

### PR DESCRIPTION
**Pull request by qcdipankar**
Added Generic Framework to onboard and run VLMs in QEff

### CODE REVIEW

File: QEfficient/transformers/models/modeling_auto.py [hunk 1]
```diff
@@ -421,6 +434,606 @@ def generate(
...
+
+    def generate_inputs(self, **kwargs):
+        generate_func = self.generate_func.get(self.model.config.architectures[0])
+        if not generate_func:
+            raise ValueError(f"Input generation function for model {self.model.config.architectures} not found.")
+
+        return generate_func(**kwargs)
+
+    def generate_inputs_intern(self, **kwargs):
+        bs: int = constants.ONNX_EXPORT_EXAMPLE_BATCH_SIZE
```
Comment (quic-amitraj) [inline QEfficient/transformers/models/modeling_auto.py:566]: Please move this inside the modeling file.

File: QEfficient/transformers/models/modeling_auto.py [hunk 2]
```diff
@@ -421,6 +434,606 @@ def generate(
...
+
+        # Avoid issues due to index out of range
+        inputs["position_ids"] = torch.full(inputs["position_ids"].shape, self.ctx_len - 1)
+
+        # import ipdb; ipdb.set_trace()
+
+        return inputs, dynamic_axes, output_names
+
+    def generate_inputs_llava(self, **kwargs):
+        bs: int = constants.ONNX_EXPORT_EXAMPLE_BATCH_SIZE
```
Comment (quic-amitraj) [inline QEfficient/transformers/models/modeling_auto.py:628]: Move this inside the modeling file ``modelling_llava.py``

File: QEfficient/transformers/models/modeling_auto.py [hunk 3]
```diff
@@ -421,6 +434,606 @@ def generate(
...
+        #     # assert num_speculative_tokens cfg is acceptable if defined
+        #     if num_speculative_tokens is None:
+        #         raise TypeError("missing required argument `num_speculative_tokens` as `is_tlm` is True.")
+        #     if not isinstance(num_speculative_tokens, int) and num_speculative_tokens < 2:
+        #         ValueError(
+        #             f"`num_speculative_tokens` arg should be an integer greater than 1, got {num_speculative_tokens}"
+        #         )
+        #     num_logits_to_keep = num_speculative_tokens + 1
+        #     if prefill_seq_len < num_logits_to_keep:
+        #         raise ValueError(
```
Comment (quic-amitraj) [inline QEfficient/transformers/models/modeling_auto.py:747]: Remove commented lines.

File: QEfficient/transformers/models/modeling_auto.py [hunk 4]
```diff
@@ -421,6 +434,606 @@ def generate(
...
+        )
+        input_len = inputs["attention_mask"].sum(1, keepdims=True)
+        padded_len = inputs["input_ids"].shape[1]
+        # input_len=padded_len
+        num_chunks = -(padded_len // -prefill_seq_len)  # ceil divide without float
+        padded_len = num_chunks * prefill_seq_len  # Convert to a multiple of prompt_len
+        generation_len = self.ctx_len - input_len.max()  # in standalone this is tensor
+        assert generation_len > 0, "generation length should be greater than zero"
+        generated_ids = np.full((batch_size, generation_len + 1), self.processor.tokenizer.pad_token_id)
+        # inputs["input_ids"]=torch.nn.functional.pad(inputs["input_ids"],(0,self.seq_len_constant-inputs["input_ids"].size(1)),"constant",self.pad_token_id)
```
Comment (quic-amitraj) [inline QEfficient/transformers/models/modeling_auto.py:920]: Remove.

File: QEfficient/utils/constants.py [hunk 5]
```diff
@@ -57,12 +57,21 @@ class Constants:
...
+    SEQ_LEN_VLM = 3072
+    CTX_LEN_VLM_INTERN = 4096
     PROMPT_LEN = 8
     INPUT_STR = ["My name is"]
     GB = 2**30
     MAX_QPC_LIMIT = 30
     MAX_RETRIES = 5  # This constant will be used set the maximum number of retry attempts for downloading a model using huggingface_hub snapshot_download
     NUM_SPECULATIVE_TOKENS = 2
+    CTX_LEN_VLM_LLAVA = 1280
+    IMG_SIZE = 336
```
Comment (quic-amitraj) [inline QEfficient/utils/constants.py:69]: Are you using these at the time of export to define shapes?

File: tests/transformers/models/test_image_text_to_text.py [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+    n_layer_text = 1
+    n_layer_vision = 1
+
+    check_vlm_pytorch_vs_kv_vs_ort_vs_ai100(
+        model_name=model_name, n_layer_text=n_layer_text, n_layer_vision=n_layer_vision
+    )
+
+
+# if __name__ == "__main__":
+#     # model_name = "OpenGVLab/InternVL2_5-1B"
```
Comment (quic-amitraj) [inline tests/transformers/models/test_image_text_to_text.py:364]: Remove commented parts.

File: QEfficient/base/modeling_qeff.py [hunk 7]
```diff
@@ -251,6 +252,7 @@ def _compile(
         if num_speculative_tokens:
             compile_hash.update(to_hashable({"num_speculative_tokens": num_speculative_tokens}))
 
+        # import ipdb; ipdb.set_trace()
```
Comment (quic-amitraj) [inline QEfficient/base/modeling_qeff.py:255]: Remove these lines.

File: QEfficient/base/pytorch_transforms.py [hunk 8]
```diff
@@ -40,12 +40,21 @@ class ModuleMappingTransform(PytorchTransform):
...
         transformed = False
         for module in model.modules():
+            # if repl_module := cls._module_mapping.get(type(module)):
+            if repl_module := cls._module_mapping.get(module.__class__.__name__):
+                module.__class__ = repl_module
+                # Handling the __init__ calls in the models
+                if hasattr(module, "__qeff_init__"):
+                    module.__qeff_init__()
+                transformed = True
+
```
Comment (quic-amitraj) [inline QEfficient/base/pytorch_transforms.py:50]: Can we combine both if conditions?

File: QEfficient/transformers/models/modeling_auto.py [hunk 9]
```diff
@@ -421,6 +434,606 @@ def generate(
...
+        # inputs["position_ids"]=torch.nn.functional.pad(inputs["position_ids"],(0,self.seq_len_constant-inputs["position_ids"].size(1)),"constant",-1)
+        input_ids = inputs["input_ids"]
+        input_ids_size = input_ids.shape[1]
+        # attention_mask = inputs["attention_mask"]
+        inputs["input_ids"] = torch.nn.functional.pad(
+            inputs["input_ids"], (0, 3072 - input_ids_size), "constant", self.processor.tokenizer.pad_token_id
```
Comment (quic-akuruvil) [inline QEfficient/transformers/models/modeling_auto.py:926]: Please avoid hardcoded value.

Comment (quic-akuruvil) [reply QEfficient/transformers/models/modeling_auto.py:926]: Make this value generic, and fetch from qpc session - `prefill_seq_len`. For whichever value it was compiled for.

File: QEfficient/transformers/models/modeling_auto.py [hunk 10]
```diff
@@ -421,6 +434,606 @@ def generate(
...
+            self.num_layers = model.config.llm_config.num_hidden_layers
+            self.num_key_value_heads = model.config.llm_config.num_key_value_heads
+            self.head_dim = model.config.llm_config.hidden_size // model.config.llm_config.num_attention_heads
+            self.pad_token_id = model.config.llm_config.pad_token_id
+            self.ctx_len = Constants.CTX_LEN_VLM_INTERN
+        elif model.config.architectures[0] == "LlavaForConditionalGeneration":
+            breakpoint()
+            self.model.config.use_cache = True
+            self.processor = processor
+            self.num_layers = model.config.text_config.num_hidden_layers
```
Comment (quic-akuruvil) [inline QEfficient/transformers/models/modeling_auto.py:495]: Make fetching `num_layers` generic. Also the padding shape. Please refer llava PR, use a similar function, which fetches based on model architecture.

**Comment by quic-amitraj** [comment]
Already addressed in #267.
